File: image-processing/image-processing.c
#include <stdio.h>
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <sstream>
using namespace std;

int main() {

	string line;
	ifstream imageFile("straight.pgm");

	int leftScore = 0;
	int rightScore = 0;
	int width = 640;

	if (imageFile.is_open()) {
		// Get the first line and do nothing
		getline(imageFile, line);

		// Get the second line which has the dimensions
		getline(imageFile, line);

		// Get the third line which has the maximum value, do nothing with it
		getline(imageFile, line);

		int i = 0;

		while (getline(imageFile, line, ' ')) {

			if (i>(width-1)) {
				i=0;
			}

			istringstream buffer(line);
			int value;
			buffer >> value;

			if (i<(width/2)) {
				leftScore += value;
			} else {
				rightScore += value;
			}

			i++;
		}

		imageFile.close();
	} else {
		cout << "Unable to open file";
	}

	cout << "Left score: " << leftScore << "\n";
	cout << "Right score: " << rightScore << "\n";

	float delta = 1.05;

	if (leftScore > rightScore) {
		float score = (float)leftScore/(float)rightScore;
		if (score > delta) {
			cout << "Turn Right\n";
		} else {
			cout << "Go Straight\n";
		}
	} else {
		float score = (float)rightScore/(float)leftScore;
		if (score > delta) {
			cout << "Turn Left\n";
		} else {
			cout << "Go Straight\n";
		}
	}

	return(0);
}

File: image-processing/image-processing.py
import sys, getopt, Image, io, urllib

def main(argv):
    inputfile = ''
    width = ''
    arguments = False
    debug = False;
    try:
        opts, args = getopt.getopt(argv, "i:d")
    except getopt.GetoptError:
        print 'image-processing.py -i <inputfile> -d <debug>'
        sys.exit(2)

    for opt, arg in opts:
        if opt == '-i':
            inputfile = arg
            arguments = True
        if opt == '-d':
            debug = True
        
    if (arguments) :
        image_file = Image.open(inputfile)
    else:
        url = urllib.urlopen("http://localhost:8080/?action=snapshot")
        snapshot = io.BytesIO(url.read())
        image_file = Image.open(snapshot)

    image_file = image_file.convert('1') # convert image to black and white
    (width,height) = image_file.size
    half = (width/2) -1

    leftscore = 0
    rightscore = 0

    for i in range(0, height-2):
        for j in range(0, half):
            leftscore += image_file.getpixel((j,i))
        for j in range(half+1, width-1):
            rightscore += image_file.getpixel((j,i))

    if (debug) :
        print("Left Score: " + str(leftscore))
        print("Right Score: " + str(rightscore))

    delta = 1.05

    if leftscore > rightscore:
        score = float(leftscore) / float(rightscore)
        if (debug) :
            print score
        if score > delta:
            print("RIGHT")
        else:
            print("STRAIGHT")
    else:
        score = float(rightscore) / float(leftscore)
        if (debug) :
            print score
        if score > delta:
            print("LEFT")
        else:
            print("STRAIGHT")

if __name__ == "__main__":
   main(sys.argv[1:])
        
